hnn_core/lfp.py

```diff
@@ -0,0 +1,237 @@
+"""
+LFPsim - Simulation scripts to compute Local Field Potentials (LFP) from cable
+compartmental models of neurons and networks implemented in NEURON simulation
+environment.
+
+LFPsim works reliably on biophysically detailed multi-compartmental neurons
+with ion channels in some or all compartments.
+
+Last updated 12-March-2016
+Developed by : 
+Harilal Parasuram & Shyam Diwakar
+Computational Neuroscience & Neurophysiology Lab, School of Biotechnology, 
+Amrita University, India.
+
+Email: harilalp@am.amrita.edu; shyam@amrita.edu
+www.amrita.edu/compneuro 
+
+translated to Python and modified to use use_fast_imem by Sam Neymotin
+based on mhines code
+
+References
+----------
+1 . Parasuram H, Nair B, D'Angelo E, Hines M, Naldi G, Diwakar S (2016)
+Computational Modeling of Single Neuron Extracellular Electric Potentials
+and Network Local Field Potentials using LFPsim. Front Comput Neurosci 10:65
+[PubMed].
+"""
+
+from neuron import h
+
+import matplotlib.pyplot as plt
+
+
+# get all Sections
+def getallSections(ty='Pyr'):
+    ls = h.allsec()
+    ls = [s for s in ls if s.name().count(ty) > 0 or len(ty) == 0]
+    return ls
+
+
+class LFPElectrode:
+    """LFP electrode class.
+
+    Parameters
+    ----------
+    coord : tuple
+        The (x, y, z) coordinates of the LFP electrode.
+    sigma : float
+        Extracellular conductivity in mS/cm (uniform for simplicity)
+    pc : instance of h.ParallelContext()
+        ParallelContext instance for running in parallel
+    method : str
+        'psa' (default), i.e., point source approximation or line source
+        approximation, i.e., 'lsa'
+
+    Attributes
+    ----------
+    lfp_t : instance of h.Vector
+        The LFP time instances.
+    lfp_v : instance of h.Vector
+        The LFP voltage.
+    imem_vec : instance of h.Vector
+        The transmembrane ionic current.
+    """
+
+    def __init__(self, coord, sigma=3.0, pc=None, method='psa'):
+
+        # see http://jn.physiology.org/content/104/6/3388.long shows table of 
+        # values with conductivity
+        self.sigma = sigma
+        self.coord = coord
+        self.vres = None
+        self.vx = None
+
+        self.imem_ptrvec = None
+        self.imem_vec = None
+        self.rx = None
+        self.bscallback = None
+        self.fih = None
+        self.method = method
+
+        if pc is None:
+            self.pc = h.ParallelContext()
+        else:
+            self.pc = pc
+
+    def setup(self):
+        """Enables fast calculation of transmembrane current (nA) at
+           each segment."""
+        h.cvode.use_fast_imem(1)
+        self.bscallback = h.cvode.extra_scatter_gather(0, self.callback)
```

Comment:
I had to replace the line [here](https://github.com/jonescompneurolab/hnn/blob/master/lfp.py#L117) with this because my version of `Neuron` doesn't have the function. Maybe @samnemo do you know what's the right thing to do here?

Reply:
@jasmainak I'm guessing you saw this post?
https://www.neuron.yale.edu/phpBB/viewtopic.php?p=15713#p15713

`cvode.extra_scatter_gather` seems to replace the `cvode.beforestep_py` functionality? Anyway, I can't find it in the neuron documentation anymore.

Reply:
yeah, I figured from the API documentation that this must be the function :)